File: sample.cpp
// ── C++ SAMPLE — Press Ctrl+Shift+X to analyze ──
#include <vector>
#include <unordered_map>
#include <algorithm>

// O(1)
int getFirst(std::vector<int>& arr) {
    return arr[0];
}

// O(n)
int findMax(std::vector<int>& arr) {
    int max = arr[0];
    for (auto x : arr) {
        if (x > max) max = x;
    }
    return max;
}

// O(n^2) — nested loops
void bubbleSort(std::vector<int>& arr) {
    for (int i = 0; i < arr.size(); i++) {
        for (int j = 0; j < arr.size() - i - 1; j++) {
            if (arr[j] > arr[j + 1])
                std::swap(arr[j], arr[j + 1]);
        }
    }
}

// O(2^n) — recursion
int fibonacci(int n) {
    if (n <= 1) return n;
    return fibonacci(n - 1) + fibonacci(n - 2);
}

// O(n) with unordered_map
std::vector<int> twoSum(std::vector<int>& nums, int target) {
    unordered_map<int, int> map;
    for (int i = 0; i < nums.size(); i++) {
        int complement = target - nums[i];
        if (map.count(complement)) return {map[complement], i};
        map[nums[i]] = i;
    }
    return {};
}

// O(n log n)
void sortArray(std::vector<int>& arr) {
    std::sort(arr.begin(), arr.end());
}

File: sample.java
// ── JAVA SAMPLE — Press Ctrl+Shift+X to analyze ──
import java.util.HashMap;
import java.util.Arrays;

public class sample {

    // O(1)
    public int getFirst(int[] arr) {
        return arr[0];
    }

    // O(n)
    public int findMax(int[] arr) {
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) max = arr[i];
        }
       
        return max;
    }

    // O(n^2) — nested loops
    public void bubbleSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }

    // O(2^n) — recursion
    public int fibonacci(int n) {
        if (n <= 1) return n;
        return fibonacci(n - 1) + fibonacci(n - 2);
    }

    // O(n) with HashMap
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) return new int[]{map.get(complement), i};
            map.put(nums[i], i);
        }
        return new int[]{};
    }

    // O(n log n)
    public void sortArray(int[] arr) {
        Arrays.sort(arr);
    }
    public int binarySearch(int arr[],int target)
    {
        int low=0,high=arr.length-1,mid;
        while(low<=high)
        {
            mid=(low+(high-low))/2;
            if(arr[mid]==target)
                return mid;
            else if(arr[mid]<target)
                low=mid+1;
            else
                high=mid-1;
        }
        return -1;
    }
}
